ggu/main.c

```diff
@@ -16,11 +16,313 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <stdbool.h>
+
+#include <rte_ip.h>
+#include <rte_udp.h>
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_malloc.h>
+#include <rte_ethdev.h>
+
+#include "gatekeeper_gk.h"
 #include "gatekeeper_ggu.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
+#include "gatekeeper_config.h"
+
+/* The UDP source and destination port numbers for GGU. */
+#define GGU_SRC_PORT (0xA0A0)
+#define GGU_DST_PORT (0XB0B0)
+
+static int
+policy_parse_next(uint16_t *proto_ptr, uint8_t *policy_ptr,
+	struct ggu_policy *policy)
+{
+	if (proto_ptr == NULL || policy_ptr == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: invalid pointor when parsing policies at %s\n", __func__);
+		RTE_ASSERT(false);
+		return -1;
+	}
+
+	policy->flow.proto = rte_be_to_cpu_16(*(uint16_t *)proto_ptr);
+	proto_ptr = proto_ptr + 1;
+
+	if (policy->flow.proto == ETHER_TYPE_IPv4) {
+		rte_memcpy(&policy->flow.f.v4, policy_ptr,
+				sizeof(policy->flow.f.v4));
+		policy_ptr = policy_ptr + sizeof(policy->flow.f.v4);
+	} else if (policy->flow.proto == ETHER_TYPE_IPv6) {
+		rte_memcpy(&policy->flow.f.v6, policy_ptr,
+				sizeof(policy->flow.f.v6));
+		policy_ptr = policy_ptr + sizeof(policy->flow.f.v6);
+	} else {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: parse policy with unknown network layer protocol %hhu\n",
+			policy->flow.proto);
+		return -1;
+	}
+
+	rte_memcpy(&policy->params, policy_ptr, sizeof(policy->params));
+	policy_ptr = policy_ptr + sizeof(policy->params);
+
+	return 0;
+}
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Calculate the RSS hash value for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	uint32_t rss_hash_val = rss_ip_flow_hf(&policy->flow, 0, 0);
+
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = -1;
+	uint32_t i;
+	uint32_t idx;
+	uint32_t shift;
+	uint16_t queue_id;
+	struct gk_config *gk_conf = get_gk_conf();
+	struct gk_cmd_entry entry;
+
+	/*
+	 * XXX Change the mapping rss hash value to rss reta entry
+	 * if the reta size is not 128.
+	 */
+	RTE_ASSERT(ggu_conf->rss_conf.reta_size == 128);
+	rss_hash_val = (rss_hash_val & 127);
+
+	idx = rss_hash_val / RTE_RETA_GROUP_SIZE;
+	shift = rss_hash_val % RTE_RETA_GROUP_SIZE;
+	queue_id = ggu_conf->rss_conf.reta_conf[idx].reta[shift];
+
+	/* XXX Change mapping queue id to the gk instance id efficiently. */
+	for (i = gk_conf->lcore_start_id; i<= gk_conf->lcore_end_id; i++)
```

Comment:
In this instance, since the for loop spans multiple lines, I think it's better to put braces around it (and it's OK, according to the style guidelines).